[LV][TTI] Calculate cost of extracting last index in a scalable vector

david-arm opened the pull request:
There are a couple of places in the loop vectoriser where we
want to calculate the cost of extracting the last lane in a
vector. However, we wrongly assume that asking for the cost
of extracting lane (VF.getKnownMinValue() - 1) is an accurate
representation of the cost of extracting the last lane. For
SVE at least, this is non-trivial as it requires the use of
whilelo and lastb instructions.

To solve this problem I have added a new
getReverseVectorInstrCost interface where the index is used
in reverse from the end of the vector. Suppose a vector has
a given ElementCount EC, the extracted/inserted lane would be
EC - 1 - Index. For scalable vectors this index is unknown at
compile time. I've added a AArch64 hook that better represents
the cost, and also a RISCV hook that maintains compatibility
with the behaviour prior to this PR.

I've also taken the liberty of adding support in vplan for
calculating the cost of VPInstruction::ExtractLastElement.

fhahn commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1465,14 +1465,21 @@ class TargetTransformInfo {
 
+  enum : int {
+    UnknownIndex = -1,
+    LastIndex = -2,
```
Given that we have `isKnownVectorIndex` below that returns true for positive indices, it seems it would be more consistent to not use LastIndex for fixed-width vectors; in those cases the last index would be known. 

AFAICT we could keep the index unsigned, and still have -1/-2 as special values? If we keep it as `int` would be good to be clear what other negative values mean.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5342,17 +5342,16 @@ LoopVectorizationCostModel::getUniformMemOpCost(Instruction *I,
+           : TTI.getVectorInstrCost(
+                 Instruction::ExtractElement, VectorTy, CostKind,
+                 VF.isScalable() ? TargetTransformInfo::LastIndex
+                                 : VF.getKnownMinValue() - 1));
```
This should now work?
```suggestion
                                 : VF.getFixedValue() - 1));
```

fhahn commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1486,7 +1493,7 @@ class TargetTransformInfo {
   /// 'Idx' being the extract lane.
   LLVM_ABI InstructionCost getVectorInstrCost(
-      unsigned Opcode, Type *Val, TTI::TargetCostKind CostKind, unsigned Index,
+      unsigned Opcode, Type *Val, TTI::TargetCostKind CostKind, int Index,
```
Do we need to update the TTI hooks to check if LastIndex has been passed, to avoid unexpected results now that we are calling it with a potentially unexpected value -2?

david-arm replied:
I have done that implicitly by changing all the TTI hooks to call `isKnownVectorIndex` when testing if the index is known or not. Any code that previously compared against -1 now calls `isKnownVectorIndex`, which returns false whether the index is -1 or -2 so as far as I can tell the change is NFC for all other targets.

There may be places where target hooks were never checking for -1 and always assuming a known index - I figured in those cases my patch doesn't make things any more broken than they were already.

fhahn replied:
Ah yes I see. For those places, could we assert that we either have `UnknownIndex` or a known index, i.e. no LastIndex, except for RISV which also supports scalable vectors and would be queried with LastIndex?

paulwalker-arm commented:
Given my faux par, as a counter recommendation what about splitting the index in two.  Pass the index as an  ElementCount and have another parameter to set the direction.  This has the advantages I mention above and for the last index you'd use an index of 0 with a direction paramater of back/tail/reverse etc.

davemgreen reviewed (commented):
getReverseVectorInstrCost sounds like it would cost llvm.vector.reverse instructions. What do you think about getVectorInstrCostFromEnd or something like it?

fhahn reviewed (commented):
> getReverseVectorInstrCost sounds like it would cost llvm.vector.reverse instructions. What do you think about getVectorInstrCostFromEnd or something like it?

This is only for vector extracts/inserts from the end, right? It think it might be good to be explcity about that in the name, although  `getVectorInstrCost`  similarly only applies to insert/extracts....

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5313,13 +5313,18 @@ LoopVectorizationCostModel::getUniformMemOpCost(Instruction *I,
+  InstructionCost Cost =
+      TTI.getAddressComputationCost(PtrTy) +
+      TTI.getMemoryOpCost(Instruction::Store, ValTy, Alignment, AS, CostKind);
+  if (!IsLoopInvariantStoreValue) {
+    if (VF.isFixed())
+      Cost += TTI.getVectorInstrCost(Instruction::ExtractElement, VectorTy,
+                                     CostKind, VF.getKnownMinValue() - 1);
+    else
+      Cost += TTI.getReverseVectorInstrCost(Instruction::ExtractElement,
+                                            VectorTy, CostKind, 0);
+  }
```
The reason for the new TTI function is so you don't need to write code like this but can instead write code that is independent of the vector type.

It would be better if for fixed length types the base implementation of the new function simply calls the existing one with a modified index.

davemgreen reviewed (approved):
The cost and AArch64 parts look OK to me. (But I'm not married to the name if someone has a better suggestion).

paulwalker-arm reviewed (approved):
A couple of suggestions but otherwise this look good to me.

david-arm committed:
[LV][TTI] Calculate cost of extracting last index in a scalable vector

There are a couple of places in the loop vectoriser where we
want to calculate the cost of extracting the last lane in a
vector. However, we wrongly assume that asking for the cost
of extracting lane (VF.getKnownMinValue() - 1) is an accurate
representation of the cost of extracting the last lane. For
SVE at least, this is non-trivial as it requires the use of
whilelo and lastb instructions.

To solve this problem I have added a new
getReverseVectorInstrCost interface where the index is used
in reverse from the end of the vector. Suppose a vector has
a given ElementCount EC, the extracted/inserted lane would be
EC - 1 - Index. For scalable vectors this index is unknown at
compile time. I've added a AArch64 hook that better represents
the cost, and also a RISCV hook that maintains compatibility
with the behaviour prior to this PR.

I've also taken the liberty of adding support in vplan for
calculating the cost of VPInstruction::ExtractLastElement.

david-arm committed:
Address review comments
